src/query.rs

```diff
@@ -78,35 +90,91 @@ impl IntoResult for QueryExecutionResult {
     }
 }
 
-/// Take an EDN query string, a reference to an open SQLite connection, a Mentat schema, and an
-/// optional collection of input bindings (which should be keyed by `"?varname"`), and execute the
-/// query immediately, blocking the current thread.
-/// Returns a structure that corresponds to the kind of input query, populated with `TypedValue`
-/// instances.
-/// The caller is responsible for ensuring that the SQLite connection has an open transaction if
-/// isolation is required.
-pub fn q_once<'sqlite, 'schema, 'query, T>
+fn fetch_values<'sqlite, 'schema>
 (sqlite: &'sqlite rusqlite::Connection,
  schema: &'schema Schema,
- query: &'query str,
- inputs: T) -> QueryExecutionResult
-        where T: Into<Option<QueryInputs>>
-{
-    let parsed = parse_find_string(query)?;
-    let algebrized = algebrize_with_inputs(schema, parsed, 0, inputs.into().unwrap_or(QueryInputs::default()))?;
+ entity: Entid,
+ attribute: Entid,
+ only_one: bool) -> QueryExecutionResult {
+    let v = Variable::from_valid_name("?v");
+
+    // This should never fail.
+    // TODO: it should be possible to algebrize with variable entity and attribute,
+    // particularly with known type, allowing the use of prepared statements.
+    let pattern = Pattern::simple(PatternNonValuePlace::Entid(entity),
+                                  PatternNonValuePlace::Entid(attribute),
+                                  PatternValuePlace::Variable(v.clone()))
+                        .unwrap();
+
+    let element = Element::Variable(v);
+    let spec = if only_one { FindSpec::FindScalar(element) } else { FindSpec::FindColl(element) };
+    let query = FindQuery::simple(spec,
+                                  vec![WhereClause::Pattern(pattern)]);
+
+    let algebrized = algebrize_with_inputs(schema, query, 0, QueryInputs::default())?;
+
+    run_algebrized_query(sqlite, algebrized)
+}
+
+fn lookup_attribute(schema: &Schema, attribute: &NamespacedKeyword) -> Result<Entid> {
+    schema.get_entid(attribute)
+          .ok_or_else(|| ErrorKind::UnknownAttribute(attribute.clone()).into())
+}
+
+/// Return a single value for the provided entity and attribute.
+/// If the attribute is multi-valued, an arbitrary value is returned.
+/// If no value is present for that entity, `None` is returned.
+/// If `attribute` isn't an attribute, `None` is returned.
+pub fn lookup_value<'sqlite, 'schema>
+(sqlite: &'sqlite rusqlite::Connection,
+ schema: &'schema Schema,
+ entity: Entid,
+ attribute: Entid) -> Result<Option<TypedValue>> {
+    fetch_values(sqlite, schema, entity, attribute, true).into_scalar_result()
```

Comment:
Correct. A _scalar_ is a single element of a field; multiple scalars are called a _vector_.

You'll see that there are four kinds of `QueryResults` — scalar, coll, tuple, rel.

(We (and Datomic) happen to not call `tuple or `coll` "`vector`" because the overlap with the data type is ambiguous.)